src/chainstate/coordinator/mod.rs

```diff
@@ -845,12 +1196,156 @@ impl<
                         );
                     }
 
+                    // compute and store information relating to exiting at a reward cycle
+                    let current_block_height = block_receipt.header.burn_header_height as u64;
+                    let current_reward_cycle = self
+                        .burnchain
+                        .block_height_to_reward_cycle(current_block_height)
+                        .ok_or_else(|| DBError::NotFoundError)?;
+                    let mut current_exit_at_rc = None;
+                    let mut current_proposal = None;
+                    let mut invalid_reward_cycles = vec![];
+                    let mut stacks_block_height_in_cycle = 1;
+                    let stacks_block_id =
+                        StacksBlockId::new(&block_receipt.header.consensus_hash, &block_hash);
+                    let parent_block_id = self
+                        .chain_state_db
+                        .get_parent(&stacks_block_id)
+                        .expect("BUG: failed to get parent for processed block");
+
+                    // look up parent in exit_at_reward_cycle_info table
+                    let exit_info_opt = SortitionDB::get_exit_at_reward_cycle_info(
+                        self.sortition_db.conn(),
+                        &parent_block_id,
+                    )?;
+
+                    if let Some(parent_exit_info) = exit_info_opt {
+                        info!("RCPR: cc: parent block info is non-None");
+                        // copy info from parent block
+                        current_exit_at_rc = parent_exit_info.curr_exit_at_reward_cycle;
+                        if parent_exit_info.block_reward_cycle < current_reward_cycle {
+                            stacks_block_height_in_cycle = 1;
+                        } else if parent_exit_info.block_reward_cycle == current_reward_cycle {
+                            current_proposal = parent_exit_info.curr_exit_proposal;
+                            stacks_block_height_in_cycle =
+                                parent_exit_info.stacks_block_height_in_cycle + 1;
+                        } else {
+                            error!("A child block should never have a lower reward cycle than its parent");
+                            panic!();
+                        }
+                        invalid_reward_cycles = parent_exit_info
```

Comment:
Hmm, it seems that `invalid_reward_cycles` grows linearly with the chain history, unbounded?  Is there a way we can avoid this (i.e. do O(1) work per reward cycle), so it doesn't lead to a node stall over time at each reward cycle boundary?

Reply:
It doesn't, it's bounded in size. 
In the logic here, we prune the invalid reward cycles of the parent, and only keep those that are greater than the current reward cycle. 
For a cycle to get added to this list, that cycle must have been chosen as a proposal and vetoed. Since there is a maximum allowable range for the voting, the size of the list can't exceed the size of that range (which is ~30).